opm/grid/common/SubGridView.hpp

```diff
@@ -0,0 +1,274 @@
+    using IntersectionIterator = typename Grid ::Traits ::LeafIntersectionIterator;
+
+    /** \brief type of the collective communication */
+    using CollectiveCommunication = typename Grid ::Traits ::CollectiveCommunication;
+
+    template <int cd>
+    struct Codim {
+        using BaseIterator = typename Grid ::Traits ::template Codim<cd>::template Partition<All_Partition>::LeafIterator;
+
+        using Entity = typename Grid ::Traits ::template Codim<cd>::Entity;
+        using EntitySeed = typename Grid ::Traits ::template Codim<cd>::EntitySeed;
+
+        using Geometry = typename Grid ::template Codim<cd>::Geometry;
+        using LocalGeometry = typename Grid ::template Codim<cd>::LocalGeometry;
+
+        /** \brief Define types needed to iterate over entities of a given partition type */
+        template <PartitionIteratorType pit>
+        struct Partition {
+            /** \brief iterator over a given codim and partition type */
+            using BaseIterator = typename Grid ::template Codim<cd>::template Partition<pit>::LeafIterator;
+        };
+    };
+
+    enum { conforming = Capabilities ::isLeafwiseConforming<Grid>::v };
+};
+
+
+template <class GridImp>
+class SubGridView
+{
+    using ThisType = SubGridView<GridImp>;
+
+public:
+    using Traits = SubGridViewTraits<GridImp>;
+
+    /** \brief type of the grid */
+    using Grid = typename Traits::Grid;
+
+    /** \brief type of the index set */
+    using IndexSet = typename Traits ::IndexSet;
+
+    /** \brief type of the intersection */
+    using Intersection = typename Traits ::Intersection;
+
+    /** \brief type of the intersection iterator */
+    using IntersectionIterator = typename Traits ::IntersectionIterator;
+
+    /** \brief type of the collective communication */
+    using CollectiveCommunication = typename Traits ::CollectiveCommunication;
+
+    /** \brief Codim Structure */
+    template <int cd>
+    struct Codim : public Traits ::template Codim<cd> {
+        using Entity = typename Traits::template Codim<cd>::Entity;
+        class Iterator
+        {
+        public:
+            Iterator(const SubGridView& view, std::size_t index)
+                : view_(&view)
+                , index_(index)
+            {
+            }
+            const Entity& operator*() const
+            {
+                entity_ = this->view_->get(index_);
+                return entity_;
+            }
+            const Entity* operator->() const
+            {
+                entity_ = this->view_->get(index_);
+                return &entity_;
+            }
+            Iterator operator++()
+            {
+                ++index_;
+                return *this;
+            }
+            Iterator operator++(int)
+            {
+                Iterator copy(*this);
+                ++index_;
+                return copy;
+            }
+            bool operator==(const Iterator& other) const
+            {
+                assert(view_ == other.view_);
```

Comment:
I think the assert here should simply be incorporated into the equality comparison. They should be cheap, anyway.

Reply:
At first I agreed, but after thinking it over I have come to disagree. Comparing iterators from two different views is a programmer error, and not a normal part of the comparison. Therefore would not the assert() be more appropriate?